Clean up trait bounds

echevrier opened the pull request:
Fixes: https://github.com/scs/substrate-api-client/issues/406
The different types that are important to communicate with a specific node can be configured.
- The configuration is similar to subxt (https://github.com/paritytech/subxt/blob/92ace0629a63fadeebd428c5b1356ba60eeabfeb/subxt/src/config.rs)
- The configuration is generic to the block type, as this requires the call which is only known from the runtime
- The API structure is now generic to the configuration and no longer to the individual types, required to communicate to the node.
    
All examples use the substrate kitchensink runtime: the config is a substrate config with the kitchensink runtime block type

Issue https://github.com/scs/substrate-api-client/issues/504 is obsolete. The ExtrinsicSigner does no more depend on Runtime input but from the configuration. It is cloneable. The StaticExtrinsicSigner is still available but instead of StaticExtrinsicSigner, we can use ExtrinsicSigner::<SubstrateConfig<Block>, sr25519::Pair>.

echevrier commented on primitives/src/extrinsics/signer.rs:
Is StaticExtrinsicSigner still needed?

echevrier replied:
The `StaticExtrinsicSigner` was added to solve #461

echevrier commented on primitives/src/extrinsics/signer.rs:
```diff
@@ -166,27 +166,16 @@ mod tests {
-
 	#[test]
-	fn test_static_extrinsic_signer_clone() {
+	fn test_extrinsic_signer_clone() {
```
Can we close #504 ?

haerdib commented on testing/examples/author_tests.rs:
```diff
@@ -15,29 +15,36 @@
 };
 
-type ExtrinsicSigner = GenericExtrinsicSigner<Pair, Signature, Runtime>;
+// This example run against a specific node.
```
You don't need to add these comments to the testing files. They are meant for CI only, not for users. Rather, the comments may get outdated and rather confuse the developer.

haerdib commented:
Sorry, I was so immersed in the code, that I totally forgot to mention: Thanks a lot for this PR! I think it's a huge improvement! 🥳

echevrier commented on primitives/src/config/substrate_config.rs:
```diff
@@ -0,0 +1,252 @@
+
+/// A struct representing the signed extra and additional parameters required
+/// to construct a transaction for the default substrate node.
+pub type SubstrateExtrinsicParams<T> = GenericExtrinsicParams<T, AssetTip<u128>>;
```
The Tip is only one configured type, so I don't think it should be in the file name. The types are configured to run with a Substrate node or a Polkadot node.  Furthermore the files are from subxt, changing the names is confusing. So I would prefer to rename them like in subxt: substrate.rs and polkadot.rs.

haerdib replied:
I meant the type name (but file name as well I guess), but for me the question is: What is a Substrate node? Substrate has a template node (plain tip) as well as the kitchensink runtime (asset tip). So for me, the name doesn't really say a thing about the difference between the two types.

I don't think we need to follow the subxt conversion here. The types are simple enough, so it's fine if we deviate over time.

echevrier replied:
I changed the types to  `AssetTipExtrinsicParams` and `PlainTipExtrinsicParams`. I also rename the file into `substrate_kitchensink.rs`

echevrier commented on primitives/src/config/substrate_config.rs:
```diff
@@ -0,0 +1,252 @@
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Encode)]
+pub struct BlakeTwo256;
+
+impl Hasher for BlakeTwo256 {
```
subxt inserted this in https://github.com/paritytech/subxt/commit/b316301d61f924c510c40f505995f28a83c1365f#diff-57540144ef1381f66537718f59e717110353e5408cdf539b67f7a1ff9cdd0d10 to make sp_core and sp_runtime dependencies optional. See https://github.com/paritytech/subxt/pull/760.
The Header trait has also been inserted for this purpose.
Removing these would certainly simplify our configurations and code. If our goal is to have an easy to use configuration, we should do it. But I like subxt idea of  "Reducing the need to keep up to date with Substrate packlage versions". See PR  https://github.com/paritytech/subxt/pull/760

haerdib replied:
Since we're not using package versions anyway, that's not really a problem for us. However, I do see the point of wasm and huge dependency tree. But in that case a separate issue is necessary because we currently do depend on sp_core and sp_runtime. Not sure how easily removable that is.

If you think it worth keeping this trait, that's fine. But it probably makes sense to move the definition to another file for better visibility. We already have two dedicated files with substrate types:
- https://github.com/scs/substrate-api-client/blob/master/primitives/src/types.rs
- https://github.com/scs/substrate-api-client/blob/master/primitives/src/serde_impls.rs

echevrier commented on primitives/src/config/substrate_config.rs:
```diff
@@ -0,0 +1,252 @@
+	type Index = u32;
+	type Hash = H256;
+	type AccountId = AccountId32;
+	type Address = MultiAddress<Self::AccountId, u32>;
```
- BlockNumber was removed in subxt:  See: https://github.com/paritytech/subxt/pull/804. Matter of taste. But, if we add it to the config, it simplifies our code like above.
- AccountIndex is not used in substrate-api-client, has no traits and was removed from node-template : https://github.com/paritytech/substrate/pull/9149. So I don't see the benefit to add it.

echevrier commented on primitives/src/config/substrate_config.rs:
```diff
@@ -0,0 +1,252 @@
+	type ExtrinsicParams = SubstrateExtrinsicParams<Self>;
+	type CryptoKey = sr25519::Pair;
+	type ExtrinsicSigner = ExtrinsicSigner<Self, Self::CryptoKey>;
+	type Block = Block;
```
The Block depends also on the ExtrinsicParams type. Actually the Extrinsic which depends on ExtrinsicParams , Adress and Signature. So to be clean, we should make the config generic on the Call.

haerdib commented on primitives/src/extrinsics/signer.rs:
```diff
@@ -91,17 +85,6 @@ where
 	}
 }
 
-impl<Signer, Runtime> From<Signer> for ExtrinsicSigner<Signer, Signer::Signature, Runtime>
```
I tested it out of curiosity. It seems to work with the code pasted below. But maybe there's even a simpler solution

```rust
impl<T, Signer> From<Signer> for ExtrinsicSigner<T>
where
	T: Config,
	Signer: Pair + Into<T::CryptoKey>,
{
	fn from(value: Signer) -> Self {
		ExtrinsicSigner::<T>::new(value.into())
	}
}
```

The removed test can also be included again with slight changes
```rust
#[test]
	fn test_extrinsic_signer_from_sr25519_pair() {
		let alice: sr25519::Pair = Pair::from_string(
			"0xe5be9a5092b81bca64be81d212e7f2f9eba183bb7a90954f7b76361f6edb5c0a",
			None,
		)
		.unwrap();

		let es_converted: ExtrinsicSigner<SubstrateKitchensinkConfig> = alice.clone().into();
		let es_new = ExtrinsicSigner::<SubstrateKitchensinkConfig>::new(alice.clone());

		assert_eq!(es_converted.signer.public(), es_new.signer.public());
	}
```

haerdib commented on primitives/src/serde_impls.rs:
```diff
@@ -458,12 +460,285 @@ mod apis_serialize {
+	type SignaturePayload = ();
+}
+
+/// A generic Substrate block type, adapted from `sp_runtime::generic::Bloc`.
```
```suggestion
/// A generic Substrate block type, adapted from `sp_runtime::generic::Block`.
```

haerdib commented on primitives/src/serde_impls.rs:
```diff
@@ -458,12 +460,285 @@ mod apis_serialize {
+	use node_template_runtime::{BalancesCall, RuntimeCall, SignedExtra};
 	use primitive_types::H256;
+	use sp_core::{crypto::Ss58Codec, Pair, H256 as Hash};
+	use sp_runtime::{generic::Era, testing::sr25519, AccountId32, MultiAddress, MultiSignature};
```
```suggestion
	use sp_runtime::{testing::sr25519, AccountId32, MultiAddress, MultiSignature};
```

haerdib commented on primitives/src/serde_impls.rs:
```diff
@@ -542,4 +817,73 @@ mod tests {
+		assert_eq!(header, header_des);
+		let extrinsics_des = block_des.extrinsics;
+		assert_eq!(extrinsics_des, extrinsics);
+	}
```
Cool, thanks a lot for these tests! That's awesome ❤️

haerdib commented on primitives/src/traits.rs:
```diff
@@ -0,0 +1,89 @@
+	fn hash(&self) -> <Self::Hasher as Hasher>::Output {
+		<<Self::Header as Header>::Hasher>::hash_of(self.header())
+	}
+	//Self::Hash {
```
What's this for?

haerdib replied:
Is this needed?

haerdib commented on src/api/api_client.rs:
```diff
@@ -301,12 +294,13 @@ where
-	use kitchensink_runtime::Runtime;
-	use sp_core::{sr25519::Pair, H256};
+	use sp_core::H256;
+	use sp_runtime::traits::GetRuntimeBlockType;
```
```suggestion
```

haerdib commented on src/api/rpc_api/author.rs:
```diff
@@ -267,18 +263,13 @@ where
 where
+	T: Config,
 	Client: Subscribe + Request,
-	Params: ExtrinsicParams<Runtime::Index, Runtime::Hash>,
```
Very nice! That's how I've imagined it. Way cleaner than previously

haerdib commented on src/api/rpc_api/events.rs:
```diff
@@ -215,14 +205,18 @@ mod tests {
+	use sp_core::{crypto::Ss58Codec, sr25519, H256};
+	use sp_runtime::{
+		generic::Block,
+		traits::{GetRuntimeBlockType, Hash},
```
```suggestion
		traits::Hash,
```

haerdib commented on examples/examples/staking_batch_payout.rs:
```diff
@@ -50,8 +52,8 @@ async fn main() {
-	let mut api = Api::<_, _, AssetTipExtrinsicParams<Runtime>, Runtime>::new(client).unwrap();
-	api.set_signer(ExtrinsicSigner::<_, Signature, Runtime>::new(alice));
+	let mut api = Api::<SubstrateKitchensinkConfig, _>::new(client).unwrap();
+	api.set_signer(ExtrinsicSigner::<SubstrateKitchensinkConfig>::new(alice.clone()));
```
```suggestion
	api.set_signer(ExtrinsicSigner::<SubstrateKitchensinkConfig>::new(alice));
```

haerdib commented on src/extrinsic/offline_extrinsic.rs:
```diff
@@ -19,22 +19,27 @@
-{
+type ExtrinsicAddress<T> =
+	<<T as Config>::ExtrinsicSigner as SignExtrinsic<<T as Config>::AccountId>>::ExtrinsicAddress;
+type Signature<T> =
```
Can't you take the Signature directly from the Config? Respective, is that type defintion actually necessary? Would it not be possible to use
```rust
// ExtrinsicAddress
T::Address
// Signture
T::Signature
```

Don't think there's a shortcut for the SignedExtra though.

echevrier replied:
Compiler requires these types' definition:

```
error[E0308]: arguments to this function are incorrect

--> src/extrinsic/offline_extrinsic.rs:44:20
   |
44 |             Some(signer) => compose_extrinsic_offline!(signer, call, self.extrinsic_params(nonce)),
   |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: expected `ac_primitives::Config::Address`, found `ac_primitives::SignExtrinsic::ExtrinsicAddress`

```

haerdib replied:
Ah, that's a pity. But yeah, makes sense. Leave as is then

haerdib commented on src/extrinsic/staking.rs:
```diff
@@ -151,20 +151,22 @@ pub trait StakingExtrinsics {
+	type Extrinsic<Call> = UncheckedExtrinsicV4<
+		Self::Address,
+		Call,
+		<T::ExtrinsicSigner as SignExtrinsic<T::AccountId>>::Signature,
```
```suggestion
		T::Signature,
```

echevrier replied:
See https://github.com/scs/substrate-api-client/pull/541#discussion_r1182404258

haerdib commented on src/extrinsic/utility.rs:
```diff
@@ -55,18 +56,16 @@ pub trait UtilityExtrinsics {
 {
 	type Extrinsic<Call> = UncheckedExtrinsicV4<
-		Signer::ExtrinsicAddress,
+		<T::ExtrinsicSigner as SignExtrinsic<T::AccountId>>::ExtrinsicAddress,
```
```suggestion
		T::Address,
```

echevrier replied:
See https://github.com/scs/substrate-api-client/pull/541#discussion_r1182404258

haerdib commented on src/extrinsic/utility.rs:
```diff
@@ -55,18 +56,16 @@ pub trait UtilityExtrinsics {
 		Call,
-		Signer::Signature,
-		Params::SignedExtra,
+		<T::ExtrinsicSigner as SignExtrinsic<T::AccountId>>::Signature,
```
```suggestion
		T::Signature,
```

echevrier replied:
See https://github.com/scs/substrate-api-client/pull/541#discussion_r1182404258

haerdib commented on testing/examples/events_tests.rs:
```diff
@@ -38,13 +38,15 @@ impl StaticEvent for ExtrinsicSuccess {
 	const EVENT: &'static str = "ExtrinsicSuccess";
 }
 
+type KitchensinkConfig = SubstrateKitchensinkConfig;
```
```suggestion

```

haerdib commented on testing/examples/events_tests.rs:
```diff
@@ -38,13 +38,15 @@ impl StaticEvent for ExtrinsicSuccess {
 	let alice_pair = AccountKeyring::Alice.pair();
-	let mut api = Api::<_, _, AssetTipExtrinsicParams<Runtime>, Runtime>::new(client).unwrap();
-	api.set_signer(ExtrinsicSigner::<_, Signature, Runtime>::new(alice_pair));
+	let mut api = Api::<KitchensinkConfig, _>::new(client).unwrap();
```
```suggestion
	let mut api = Api::<SubstrateKitchensinkConfig, _>::new(client).unwrap();
```

haerdib commented on testing/examples/events_tests.rs:
```diff
@@ -38,13 +38,15 @@ impl StaticEvent for ExtrinsicSuccess {
-	let mut api = Api::<_, _, AssetTipExtrinsicParams<Runtime>, Runtime>::new(client).unwrap();
-	api.set_signer(ExtrinsicSigner::<_, Signature, Runtime>::new(alice_pair));
+	let mut api = Api::<KitchensinkConfig, _>::new(client).unwrap();
+	api.set_signer(ExtrinsicSigner::<KitchensinkConfig>::new(alice_pair));
```
```suggestion
	api.set_signer(ExtrinsicSigner::<SubstrateKitchensinkConfig>::new(alice_pair));
```

haerdib commented on testing/examples/frame_system_tests.rs:
```diff
@@ -42,8 +43,8 @@ async fn main() {
 	let alice_pair = AccountKeyring::Alice.pair();
-	let mut api = Api::<_, _, AssetTipExtrinsicParams<Runtime>, Runtime>::new(client).unwrap();
-	api.set_signer(ExtrinsicSigner::<_, Signature, Runtime>::new(alice_pair));
+	let mut api = Api::<KitchensinkConfig, _>::new(client).unwrap();
```
```suggestion
	let mut api = Api::<SubstrateKitchensinkConfig, _>::new(client).unwrap();
```

haerdib commented on testing/examples/pallet_balances_tests.rs:
```diff
@@ -15,19 +15,17 @@
-	let alice_pair = AccountKeyring::Alice.pair();
-	let mut api = Api::<_, _, AssetTipExtrinsicParams<Runtime>, Runtime>::new(client).unwrap();
-	api.set_signer(alice_pair);
+	let api = Api::<KitchensinkConfig, _>::new(client).unwrap();
```
```suggestion
	let api = Api::<SubstrateKitchensinkConfig, _>::new(client).unwrap();
```

haerdib commented on testing/examples/pallet_transaction_payment_tests.rs:
```diff
@@ -16,22 +16,23 @@
 	Api, GetChainInfo, GetTransactionPayment,
 };
 
+type KitchensinkConfig = SubstrateKitchensinkConfig;
```
```suggestion
```

haerdib commented on testing/examples/pallet_transaction_payment_tests.rs:
```diff
@@ -16,22 +16,23 @@
 	let alice_pair = AccountKeyring::Alice.pair();
-	let mut api = Api::<_, _, AssetTipExtrinsicParams<Runtime>, Runtime>::new(client).unwrap();
-	api.set_signer(ExtrinsicSigner::<_, Signature, Runtime>::new(alice_pair));
+	let mut api = Api::<KitchensinkConfig, _>::new(client).unwrap();
```
```suggestion
	let mut api = Api::<SubstrateKitchensinkConfig, _>::new(client).unwrap();
```

haerdib commented on testing/examples/pallet_transaction_payment_tests.rs:
```diff
@@ -16,22 +16,23 @@
-	let mut api = Api::<_, _, AssetTipExtrinsicParams<Runtime>, Runtime>::new(client).unwrap();
-	api.set_signer(ExtrinsicSigner::<_, Signature, Runtime>::new(alice_pair));
+	let mut api = Api::<KitchensinkConfig, _>::new(client).unwrap();
+	api.set_signer(<KitchensinkConfig as Config>::ExtrinsicSigner::new(alice_pair));
```
```suggestion
	api.set_signer(alice_pair.into());
```

If From is implemented again

haerdib reviewed (commented):
Nice ! I really like it 🥳  Only small change requests, other than that I approve 👍

haerdib commented:
One thing: Can the Pallet Configs be remove now? E.g `FrameSystemConfig` and such.

echevrier commented on src/extrinsic/staking.rs:
```diff
@@ -151,20 +151,22 @@ pub trait StakingExtrinsics {
-	type Extrinsic<Call> =
-		UncheckedExtrinsicV4<Self::Address, Call, Signer::Signature, Params::SignedExtra>;
+	type AccountId = T::AccountId;
+	type Address = <T::ExtrinsicSigner as SignExtrinsic<T::AccountId>>::ExtrinsicAddress;
```
See https://github.com/scs/substrate-api-client/pull/541#discussion_r1182404258

echevrier commented:
> One thing: Can the Pallet Configs be removed now? E.g `FrameSystemConfig` and such.

FrameSystemConfig is still used by https://github.com/scs/substrate-api-client/blob/29b953d4972f08333a4e2909947ce860253f8043/src/api/rpc_api/events.rs#L354

echevrier commented:
> > One thing: Can the Pallet Configs be removed now? E.g `FrameSystemConfig` and such.
> 
> FrameSystemConfig is still used by
> 
> https://github.com/scs/substrate-api-client/blob/29b953d4972f08333a4e2909947ce860253f8043/src/api/rpc_api/events.rs#L354

I removed the dependencies and deleted the pallet configs

haerdib reviewed (commented):
Cool! Very nice. Only one comment left :)

haerdib reviewed (approved):
Awesome, thanks a lot 🥳
